src/runtime/indexes.ts

```diff
@@ -0,0 +1,272 @@
+    }
+    return proposal;
+  }
+
+  walkPropose(proposal:Proposal, index:any, a:ResolvedValue, b:ResolvedValue, c:ResolvedValue, n:ResolvedValue,
+              fieldB:EAVNField, fieldC:EAVNField, transaction:number, round:number):Proposal {
+    let {forFields} = proposal;
+    let bIx = index[a as ID];
+    if(!bIx) {
+      proposal.cardinality = 0;
+      return proposal;
+    }
+    if(isResolved(b)) {
+      let cIx = bIx[b];
+      if(!cIx) {
+        proposal.cardinality = 0;
+        return proposal;
+      }
+      if(isResolved(c)) {
+        let ntrcArray = cIx[c];
+        if(sumTimes(ntrcArray, transaction, round) > 0) {
+          proposal.skip = true;
+          return proposal;
+        } else {
+          proposal.cardinality = 0;
+          return proposal;
+        }
+      } else {
+        forFields[0] = fieldC;
+        proposal.info = Object.keys(cIx);
+        proposal.cardinality = proposal.info.length;
+        return proposal;
+      }
+    } else {
+      forFields[0] = fieldB;
+      proposal.info = Object.keys(bIx);
+      proposal.cardinality = proposal.info.length;
+      return proposal;
+    }
+  }
+
+  walkCheck(index:any, a:ResolvedValue, b:ResolvedValue, c:ResolvedValue, n:ResolvedValue, transaction:number, round:number):boolean {
+    let bIx = index[a as ID];
+    if(!bIx) return false;
+    if(isResolved(b)) {
+      let cIx = bIx[b];
+      if(!cIx) return false;
+      if(isResolved(c)) {
+        let ntrcArray = cIx[c];
+        return sumTimes(ntrcArray, transaction, round) > 0;
+      } else {
+        for(let key of Object.keys(cIx)) {
+          let ntrcArray = cIx[key];
+          if(sumTimes(ntrcArray, transaction, round) > 0) {
+            return true;
+          }
+        }
+        return false;
+      }
+    } else {
+      for(let key of Object.keys(bIx)) {
+        let cIx = bIx[key];
+        if(!cIx) return false;
+        if(isResolved(c)) {
+          let ntrcArray = cIx[c];
+          return sumTimes(ntrcArray, transaction, round) > 0;
+        } else {
+          for(let key of Object.keys(cIx)) {
+            let ntrcArray = cIx[key];
+            if(sumTimes(ntrcArray, transaction, round) > 0) {
+              return true;
+            }
+          }
+          return false;
+        }
+      }
+    }
+    return false;
+  }
+
+  check(e:ResolvedValue, a:ResolvedValue, v:ResolvedValue, n:ResolvedValue, transaction = Infinity, round = Infinity):boolean {
+    if(isResolved(e)) {
+      return this.walkCheck(this.eavIndex, e, a, v, n, transaction, round);
+    } else if(isResolved(a)) {
+      return this.walkCheck(this.aveIndex, a, v, e, n, transaction, round);
+    }
+    return true;
+  }
+
+  get(e:ResolvedValue, a:ResolvedValue, v:ResolvedValue, n:ResolvedValue, transaction = Infinity, round = Infinity):EAVN[] {
+    let final = createArray() as EAVN[];
+    return final;
+  }
+}
+
+
+export class MatrixIndex implements Index {
```

Comment:
Should stick a comment above this guy marking him as not being usable yet. It's pretty obvious when in this file, but from other files there'll be no indication without a doc comment. Otherwise, you can drop the export til its ready for showtime.